nmigen_boards/mercury.py

```diff
@@ -0,0 +1,226 @@
+                 Attrs(IOSTANDARD="LVTTL")),
+        Resource("user_btn", 3, Pins("3", dir="i", conn=("input", 0)),
+                 Attrs(IOSTANDARD="LVTTL")),
+        Resource("user_btn", 4, Pins("4", dir="i", conn=("input", 0)),
+                 Attrs(IOSTANDARD="LVTTL"))
+    ]
+
+    vga = [
+        Resource("vga_out", 0,
+            Subsignal("hsync", PinsN("led_0:3", dir="o")),
+            Subsignal("vsync", PinsN("led_0:4", dir="o")),
+
+            Subsignal("r", Pins("dio_0:1 dio_0:2 dio_0:3", dir="o")),
+            Subsignal("g", Pins("dio_0:4 dio_0:5 dio_0:6", dir="o")),
+            Subsignal("b", Pins("dio_0:7 clkio_0:1", dir="o")),
+            Attrs(IOSTANDARD="LVCMOS33", SLEW="FAST")
+        )
+    ]
+
+    extclk = [
+        Resource("extclk", 0, Pins("1", dir="i", conn=("clkio", 1)),
+                 Attrs(IOSTANDARD="LVCMOS33"))
+    ]
+
+    sevenseg = [
+        Resource("sevenseg", 0,
+                 Subsignal("a", PinsN("13", dir="o", conn=("gpio", 0))),
+                 Subsignal("b", PinsN("14", dir="o", conn=("gpio", 0))),
+                 Subsignal("c", PinsN("15", dir="o", conn=("gpio", 0))),
+                 Subsignal("d", PinsN("16", dir="o", conn=("gpio", 0))),
+                 Subsignal("e", PinsN("17", dir="o", conn=("gpio", 0))),
+                 Subsignal("f", PinsN("18", dir="o", conn=("gpio", 0))),
+                 Subsignal("g", PinsN("19", dir="o", conn=("gpio", 0))),
+                 Subsignal("dp", PinsN("20", dir="o", conn=("gpio", 0))),
+                 Subsignal("en", Pins("9 10 11 12", dir="o",
+                                      conn=("gpio", 0))),
+                 Attrs(IOSTANDARD="LVTTL")
+        )
+    ]
+
+    ps2 = [
+        Resource("ps2", 0,
+                 Subsignal("clk", Pins("2", dir="io", conn=("led", 0))),
+                 Subsignal("data", Pins("1", dir="io", conn=("led", 0))),
+                 Attrs(IOSTANDARD="LVTTL")
+        )
+    ]
+
+    audio = [
+        Resource("audio", 0,
+                 Subsignal("l", Pins("30", dir="o", conn=("gpio", 0))),
+                 Subsignal("r", Pins("29", dir="o", conn=("gpio", 0))),
+                 Attrs(IOSTANDARD="LVTTL")
+        )
+    ]
+
+    def toolchain_program(self, products, name):
+        # https://github.com/cr1901/mercpcl
+        mercpcl = os.environ.get("MERCPCL", "mercpcl")
+        with products.extract("{}.bin".format(name)) as bitstream_filename:
+            subprocess.run([mercpcl, bitstream_filename], check=True)
+
+    # Convenience method to connect all baseboard I/O at once.
+    def connect_baseboard(self):
+        self.add_resources(self.user_sw + self.user_btn + self.vga +
+                           self.extclk + self.sevenseg + self.ps2 + self.audio)
+
+if __name__ == "__main__":
+    from ._blinky import Blinky
+    p = MercuryPlatform()
+    p.add_resources(p.leds)
+    p.build(Blinky("clk50"), do_program=True,
+            script_after_run=r"""-opt_mode AREA""")
```

Comment:
Is this necessary?

Reply:
Not at all, just for testing that `-opt_mode AREA` works. In `omigen`, this was the default for this platform unconditionally.

But I've come around to the idea that in `nmigen`, the user should add it on a build-by-build basis.